core/src/execution/datafusion/expressions/cast.rs

```diff
@@ -103,10 +126,78 @@ impl Cast {
             (DataType::LargeUtf8, DataType::Boolean) => {
                 Self::spark_cast_utf8_to_boolean::<i64>(&array, self.eval_mode)?
             }
-            _ => cast_with_options(&array, to_type, &CAST_OPTIONS)?,
+            (
+                DataType::Utf8,
+                DataType::Int8 | DataType::Int16 | DataType::Int32 | DataType::Int64,
+            ) => Self::spark_cast_string_to_integral(to_type, &array, self.eval_mode)?,
+            (
+                DataType::Dictionary(key_type, value_type),
+                DataType::Int8 | DataType::Int16 | DataType::Int32 | DataType::Int64,
+            ) if key_type.as_ref() == &DataType::Int32
```

Comment:
I've been assuming it to be so, though @viirya can give us the definitive answer

Reply:
I remember in many places in native code, we assume that dictionary keys are always Int32 type.

But I forgot that where we make such assumption. 😅 

cc @sunchao Do you remember that?

Reply:
Oh, I see. I think the assumption comes from native scan side where the Parquet dictionary indices is always of integer type so dictionary keys read from native scan is always Int32 type.

Reply:
You can check the `DictDecoder` in native scan implementation.

Reply:
Except for that any operator or expression during execution produce a dictionary with keys other than Int32 type. But for that I think it should be considered a bug for us to fix because I don't think it makes sense to change dictionary key type.

Reply:
> Oh, I see. I think the assumption comes from native scan side where the Parquet dictionary indices is always of integer type so dictionary keys read from native scan is always Int32 type.

Yes that is exactly right.

Reply:
Thanks @sunchao for confirming it.